### ISSUE TITLE
Stop adding rel=noreferrer to links that open in a new tab

### ISSUE THREAD

**Issue by wpjames** [WordPress/gutenberg#26914]
I am on WP version 5.5.3 and I have added the following code to my child theme functions.php.

```
function my_targeted_link_rel($rel_values) {
 	return 'noopener';
}
add_filter('wp_targeted_link_rel', 'my_targeted_link_rel',999);

```

However, Gutenberg is still adding noreferrer when I create a new link that opens in a new tab. Where is this coming from?

I don't think it's necessary to hard code rel attributes into the link module. Just let wp_targeted_link_rel add the necessary rel attributes when the post is saved/updated.

**Comment by SergeyBiryukov** [comment]
It's worth noting that WordPress core has removed `noreferrer` from `wp_targeted_link_rel()` and other uses in version 5.6, see https://core.trac.wordpress.org/changeset/49215 and https://core.trac.wordpress.org/ticket/49558 for more details.

The same change should probably be implemented in Gutenberg too.

**Comment by talldan** [comment]
> However, Gutenberg is still adding noreferrer when I create a new link that opens in a new tab. Where is this coming from?

@wpjames The block editor adds the 'noreferrer noopener' values for any links, but it doesn't use the value from `wp_targeted_link_rel`. 

Potentially it could do, though it may cause issues with block validation, so it might not be as simple as hooking the editor into `wp_targeted_link_rel`.

Probably a good first step would be removing `noreferrer` from the editor's hardcoded value.

**Comment by pablinos** [comment]
> Probably a good first step would be removing noreferrer from the editor's hardcoded value.

Wouldn't that also invalidate blocks? Are you thinking that we pick a new value and stick to it rather than having it change based on a global setting? I think we'd still have to create migrations, but the rendered output would be consistent.

Is there a way we can cope with block output changing based on a global setting like that? If someone changes the default thumbnail image size, for example, I could see that affecting block output. I realise this would break the concept of the `save` function being pure. Perhaps it's a special type of attribute, where the rendered content can be compared with a default value and presumed to be valid if they match.

**Comment by adamsilverstein** [comment]
I tested stopping adding `noreferrer` in https://github.com/WordPress/gutenberg/pull/26968. Tested with a few basic blocks with external links:

* Existing external links still included `noreferrer`.
* New links did not include `noreferrer` (only `noopener`).
* I did not encounter any block validation issues.

I encounter a linting error when trying to commit the changes, we will need to remove or exclude this linting check if we want to make this change:

```
error  Using target="_blank" without rel="noreferrer" is a security risk: see https://html.spec.whatwg.org/multipage/links.html#link-type-noopener  react/jsx-no-target-blank
```

**Comment by talldan** [comment]
> I encounter a linting error when trying to commit the changes, we will need to remove or exclude this linting check if we want to make this change:

@adamsilverstein Looks like we can use the `allowReferrer` option on this rule:
https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-target-blank.md#rule-options

I think an option would have to be added to the config here:
https://github.com/WordPress/gutenberg/blob/master/packages/eslint-plugin/configs/react.js#L9-L33

> Wouldn't that also invalidate blocks?

Yep, testing #26968 it looks like the file block needs a deprecation. 

The button and image block only use `noreferrer noopener` as a default value. Button and image blocks already in posts will keep the value `noreferrer noopener`, while newly added blocks will only have `noopener`. I think this is probably ok. What do others think?

The file block hardcodes `noreferrer noopener` into its save output so needs a deprecation. Other blocks are dynamic so won't have validation issues.

**Comment by isaumya** [comment]
Hi @talldan,
Any update on the matter? Also, I don't think it would be a good idea to keep the old links with `noreferer noopener` and then the new links will only have `noopener`.

The reason behind this is if there is no migration to automatically change the old links from `noreferer noopener`  to just `noopener` then people will still use codes like:

```php
add_filter( 'the_content', static function($content) { 
  $replace = array(" noreferrer" => "" ,"noreferrer " => "");
  $new_content = strtr($content, $replace);
  return $new_content;
 }, 99 );
```

To get rid of the `noreferrer` from the links and then this new fix will have no meaning as people will already be using a slow code that checks and update the content. The goal would be to make sure people don't have to use codes like these even for the existing links in the site.

**Comment by talldan** [comment]
@isaumya No update from me. I'm not assigned to this issue.

**Comment by isaumya** [comment]
Sorry. Hey, @adamsilverstein any update on this issue?

**Comment by Hastibe** [comment]
Is there a solution to this, yet? Having "noreferrer" automatically being added to links that open in new tabs is problematic.

**Comment by tomtolkien** [comment]
Is there any update on this issue? It's making using WordPress blocks for affiliate marketing very problematic. 

Also, the code mentioned earlier no longer provides a workaround for the issue. 

> ```
> add_filter( 'the_content', static function($content) { 
>   $replace = array(" noreferrer" => "" ,"noreferrer " => "");
>   $new_content = strtr($content, $replace);
>   return $new_content;
>  }, 99 );
> ```
>

**Comment by isaumya** [comment]
> Also, the code mentioned earlier no longer provides a workaround for the issue.

I am still using that code and it's working without any issue whatsoever. Here is a screenshot:

![usage of the_content filter](https://i.imgur.com/ftJ6RFX.jpeg)

And this is still working like charm for me. I will keep using this until these bugs are fixed in Gutenberg.

**Comment by vitalytennant** [comment]
via ticket [54410](https://github.com/WordPress/gutenberg/issues/54410)

Hello,

Thank you for your response.

Code doesn't work.

//Removes noreferrer from new or updated posts
function my_targeted_link_rel($rel_values) {
return 'noopener';
}
add_filter('wp_targeted_link_rel', 'my_targeted_link_rel',999);

Aforementioned github reference does not address rel=”noopener”

There needs to be a permanent solution, because anyone can add rel=”noreferrer noopener” if they'd like, not the other way around.

This can be further elaborated via https://core.trac.wordpress.org/ticket/59333#ticket

"This impacts the entire wordpress opensource, and should be addressed respectively, and enhanced with either WP plugin, and/or the next WP update."

Everyone (without going into intricacies of everything) has links on their wordpress websites, this impacts everyone.

**Comment by adamsilverstein** [comment]
Just a couple of things remaining for this which I noted in a comment on https://github.com/WordPress/gutenberg/pull/26968

**Comment by afercia** [comment]
It's worth noting that Core removed the `noopener` rel attribute as well in https://core.trac.wordpress.org/changeset/59120
Quoting:
> supported browsers have changed their security policies and no longer allow the opened link to have JavaScript access to the previous tab

At the moment, Core and the block editor aren't aligned and output inconsistent markup. I'd suggest raising priority on this issue and the associated PR https://github.com/WordPress/gutenberg/pull/26968 @WordPress/gutenberg-core

**Comment by t-hamano** [comment]
I agree with removing `noopener` and `noreferrer`, but these attributes exist in various places in the Gutenberg project.

Trying to address them all at once in a single PR would require multiple efforts, such as deprecating blocks and modifying test files, which would increase the review load.

I recommend treating this issue as a tracking issue and addressing and evaluating it incrementally.

**Comment by t-hamano** [comment]
Just to be sure, are the following three values ​​the ones that should be deleted?

- `nofollow`
- `noreferrer`
- `noopener`

**Comment by afercia** [comment]
> Just to be sure, are the following three values ​​the ones that should be deleted?

`nofollow` has an entirely different purpose and I don't think it should be touched.

**Comment by t-hamano** [comment]
Then, let's gradually remove `noreferrer` and `noopener`.

**Comment by aduth** [comment]
Something I haven't seen in this discussion is a distinction between links shown in the admin vs. links included in front-end post block content, and whether the `rel` should be different in one context vs. the other. In particular the influence of `noreferrer` in the destination site being able to know where traffic originated from. I see the value in allowing referrer to be shared for front-end post content, but for the admin interface it's not as clear to me that we'd want to share that from a privacy or security perspective. My "worst-case" scenario would be that we could be sharing URLs that expose sensitive information (e.g. secret values, or the existence of a plugin). However, it appears that we restrict this from by default via [`wp_admin_headers`](https://developer.wordpress.org/reference/functions/wp_admin_headers/) applying [`strict-origin-when-cross-origin`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Referrer-Policy#strict-origin-when-cross-origin_2) (also now a browser default), so only the origin would be shared (e.g. https://example.com and not https://example.com/wp-admin/whatever ).

So I'd still invite feedback whether there's a worthwhile distinction to be made here, but my impression is that the current suggested approach of "remove noreferrer and noopener everywhere" could be valid in the admin context as well.

**Comment by adamsilverstein** [comment]
Thanks @aduth good point - your analysis is right.

Looking at the 39 files this PR touches, they split roughly into two groups:

**Admin-UI links** (where your concern applies): the `ExternalLink` component, the More Menu items in `editor`/`edit-widgets`/`customize-widgets`, the `media-categories` "view more" links, and the editor-only links in `site-logo` and `categories` edit views. These are all rendered inside wp-admin, where the `strict-origin-when-cross-origin` Referrer-Policy (via `wp_admin_headers`, and now the browser default) limits outbound referrers to origin-only. No full-path URLs or query strings leak, so removing `noreferrer` here is safe.

**Front-end post content** (where sharing referrer is arguably desirable): the `file`, `latest-posts`, `button`, and `image` block output, the link format (`format-library/src/link/utils.js`), the phrasing-content paste reducer, and the associated test fixtures. These render on the front end where standard site Referrer-Policy applies and sharing the referrer is the expected web-platform default.

So  the admin context is already covered by the Referrer-Policy header, and the front-end behavior is what we want. I'll call this out in the PR description.
